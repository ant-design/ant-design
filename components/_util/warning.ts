import * as React from 'react';
import rcWarning, { resetWarned as rcResetWarned } from 'rc-util/lib/warning';

export function noop() {}

let deprecatedWarnList: Record<string, string[]> | null = null;

export function resetWarned() {
  deprecatedWarnList = null;
  rcResetWarned();
}

type Warning = (valid: boolean, component: string, message?: string) => void;

// eslint-disable-next-line import/no-mutable-exports
let warning: Warning = noop;
if (process.env.NODE_ENV !== 'production') {
  warning = (valid, component, message) => {
    rcWarning(valid, `[antd: ${component}] ${message}`);

    // StrictMode will inject console which will not throw warning in React 17.
    if (process.env.NODE_ENV === 'test') {
      resetWarned();
    }
  };
}

type BaseTypeWarning = (
  valid: boolean,
  /**
   * - deprecated: Some API will be removed in future but still support now.
   * - usage: Some API usage is not correct.
   * - breaking: Breaking change like API is removed.
   */
  type: 'deprecated' | 'usage' | 'breaking',
  message?: string,
) => void;

type TypeWarning = BaseTypeWarning & {
  deprecated: (valid: boolean, oldProp: string, newProp: string, message?: string) => void;
};

export interface WarningContextProps {
  /**
   * @descCN 设置警告等级，设置 `false` 时会将废弃相关信息聚合为单条信息。
   * @descEN Set the warning level. When set to `false`, discard related information will be aggregated into a single message.
   * @since 5.10.0
   */
  strict?: boolean;
}

export const WarningContext = React.createContext<WarningContextProps>({});

/**
 * This is a hook but we not named as `useWarning`
 * since this is only used in development.
 * We should always wrap this in `if (process.env.NODE_ENV !== 'production')` condition
 */
export const devUseWarning: (component: string) => TypeWarning =
  process.env.NODE_ENV !== 'production'
    ? (component) => {
        const { strict } = React.useContext(WarningContext);

        const typeWarning: TypeWarning = (valid, type, message) => {
          if (!valid) {
            if (strict === false && type === 'deprecated') {
              const existWarning = deprecatedWarnList;

              if (!deprecatedWarnList) {
                deprecatedWarnList = {};
              }

              deprecatedWarnList[component] = deprecatedWarnList[component] || [];
              if (!deprecatedWarnList[component].includes(message || '')) {
                deprecatedWarnList[component].push(message || '');
              }

              // Warning for the first time
              if (!existWarning) {
                console.warn(
                  '[antd] There exists deprecated usage in your code:',
                  deprecatedWarnList,
                );
              }
            } else {
              warning(valid, component, message);
            }
          }
        };

        typeWarning.deprecated = (valid, oldProp, newProp, message) => {
          typeWarning(
            valid,
            'deprecated',
            `\`${oldProp}\` is deprecated. Please use \`${newProp}\` instead.${
              message ? ` ${message}` : ''
            }`,
          );
        };

        return typeWarning;
      }
    : () => {
        const noopWarning: TypeWarning = () => {};

        noopWarning.deprecated = noop;

        return noopWarning;
      };

export default warning;
