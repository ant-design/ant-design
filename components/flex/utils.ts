import classNames from 'classnames';

import type { FlexProps } from './interface';

export const flexWrapValues: React.CSSProperties['flexWrap'][] = ['wrap', 'nowrap', 'wrap-reverse'];

export const justifyContentValues: React.CSSProperties['justifyContent'][] = [
  'flex-start',
  'flex-end',
  'start',
  'end',
  'center',
  'space-between',
  'space-around',
  'space-evenly',
  'stretch',
  'normal',
  'left',
  'right',
];

export const alignItemsValues: React.CSSProperties['alignItems'][] = [
  'center',
  'start',
  'end',
  'flex-start',
  'flex-end',
  'self-start',
  'self-end',
  'baseline',
  'normal',
  'stretch',
];

const genClsWrap = (prefixCls: string, props: FlexProps) => {
  const wrap = props.wrap === true ? 'wrap' : props.wrap;
  return {
    [`${prefixCls}-wrap-${wrap}`]: wrap && flexWrapValues.includes(wrap),
  };
};

const genClsAlign = (prefixCls: string, props: FlexProps) => {
  const alignCls: Record<PropertyKey, boolean> = {};
  alignItemsValues.forEach((cssKey) => {
    alignCls[`${prefixCls}-align-${cssKey}`] = props.align === cssKey;
  });
  alignCls[`${prefixCls}-align-stretch`] = !props.align && !!props.vertical;
  return alignCls;
};

const genClsJustify = (prefixCls: string, props: FlexProps) => {
  const justifyCls: Record<PropertyKey, boolean> = {};
  justifyContentValues.forEach((cssKey) => {
    justifyCls[`${prefixCls}-justify-${cssKey}`] = props.justify === cssKey;
  });
  return justifyCls;
};

function createFlexClassNames(prefixCls: string, props: FlexProps) {
  return classNames({
    ...genClsWrap(prefixCls, props),
    ...genClsAlign(prefixCls, props),
    ...genClsJustify(prefixCls, props),
  });
}

export default createFlexClassNames;
