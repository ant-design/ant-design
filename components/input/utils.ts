import type { ReactNode } from 'react';

import type { InputProps } from './Input';

export function hasPrefixSuffix(props: {
  prefix?: ReactNode;
  suffix?: ReactNode;
  allowClear?: InputProps['allowClear'];
  showCount?: InputProps['showCount'];
}) {
  return !!(props.prefix || props.suffix || props.allowClear || props.showCount);
}
