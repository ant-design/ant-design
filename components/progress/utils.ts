import { presetPrimaryColors } from '@ant-design/colors';

import type { CircleProps } from './Circle';
import type { ProgressProps } from './progress';

export function validProgress(progress?: number) {
  if (!progress || progress < 0) {
    return 0;
  }
  if (progress > 100) {
    return 100;
  }
  return progress;
}

export function getSuccessPercent({ success, successPercent }: ProgressProps) {
  let percent = successPercent;
  /** @deprecated Use `percent` instead */
  if (success && 'progress' in success) {
    percent = success.progress;
  }
  if (success && 'percent' in success) {
    percent = success.percent;
  }
  return percent;
}

export const getPercentage = ({ percent, success, successPercent }: ProgressProps) => {
  const realSuccessPercent = validProgress(getSuccessPercent({ success, successPercent }));
  return [realSuccessPercent, validProgress(validProgress(percent) - realSuccessPercent)];
};

export const getStrokeColor = ({
  success = {},
  strokeColor,
}: Partial<CircleProps>): (string | Record<PropertyKey, string>)[] => {
  const { strokeColor: successColor } = success;
  return [successColor || presetPrimaryColors.green, strokeColor || null!];
};

export const getSize = (
  size: ProgressProps['size'],
  type: ProgressProps['type'] | 'step',
  extra?: {
    steps?: number;
    strokeWidth?: number;
  },
): [number, number] => {
  let width = -1;
  let height = -1;
  if (type === 'step') {
    const steps = extra!.steps!;
    const strokeWidth = extra!.strokeWidth!;
    if (typeof size === 'string' || typeof size === 'undefined') {
      width = size === 'small' ? 2 : 14;
      height = strokeWidth ?? 8;
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = 14, height = 8] = (Array.isArray(size) ? size : [size.width, size.height]) as [
        number,
        number,
      ];
    }

    width *= steps;
  } else if (type === 'line') {
    const strokeWidth = extra?.strokeWidth;
    if (typeof size === 'string' || typeof size === 'undefined') {
      height = strokeWidth || (size === 'small' ? 6 : 8);
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = -1, height = 8] = (Array.isArray(size) ? size : [size.width, size.height]) as [
        number,
        number,
      ];
    }
  } else if (type === 'circle' || type === 'dashboard') {
    if (typeof size === 'string' || typeof size === 'undefined') {
      [width, height] = size === 'small' ? [60, 60] : [120, 120];
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else if (Array.isArray(size)) {
      width = (size[0] ?? size[1] ?? 120) as number;
      height = (size[0] ?? size[1] ?? 120) as number;
    }
  }
  return [width, height];
};
