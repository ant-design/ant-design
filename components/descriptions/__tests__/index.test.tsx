import React from 'react';
import MockDate from 'mockdate';

import Descriptions from '..';
import { resetWarned } from '../../_util/warning';
import mountTest from '../../../tests/shared/mountTest';
import { render } from '../../../tests/utils';
import ConfigProvider from '../../config-provider';

describe('Descriptions', () => {
  mountTest(Descriptions);

  const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

  afterEach(() => {
    MockDate.reset();
    errorSpy.mockReset();
  });

  afterAll(() => {
    errorSpy.mockRestore();
  });

  it('when max-width: 575px, column=1', () => {
    const wrapper = render(
      <Descriptions>
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
        <Descriptions.Item>No-Label</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.container.querySelectorAll('tr')).toHaveLength(5);
    expect(wrapper.container.querySelectorAll('.ant-descriptions-item-label')).toHaveLength(4);
    wrapper.unmount();
  });

  it('when max-width: 575px, column=2', () => {
    const wrapper = render(
      <Descriptions column={{ xs: 2 }}>
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.container.querySelectorAll('tr')).toHaveLength(2);
    wrapper.unmount();
  });

  it('when max-width: 575px, column=2, span=2', () => {
    const { container } = render(
      <Descriptions
        column={{ xs: 2 }}
        items={[
          {
            label: 'Product',
            children: 'Cloud Database',
            span: { xs: 2 },
          },
          {
            label: 'Billing',
            children: 'Prepaid',
            span: { xs: 1 },
          },
          {
            label: 'Time',
            children: '18:00:00',
            span: { xs: 1 },
          },
        ]}
      />,
    );

    expect(container.querySelectorAll('.ant-descriptions-item')[0]).toHaveAttribute('colSpan', '2');
    expect(container.querySelectorAll('.ant-descriptions-item')[1]).toHaveAttribute('colSpan', '1');
    expect(container.querySelectorAll('.ant-descriptions-item')[2]).toHaveAttribute('colSpan', '1');
  });

  it('span = filled', () => {
    const { container } = render(
      <Descriptions
        column={3}
        items={[
          { label: '0', children: '', span: 2 },
          { label: '1', children: '' },
          { label: '2', children: '' },
          { label: '3', children: '', span: 'filled' },
          { label: '4', children: '', span: 'filled' },
          { label: '5', children: '' },
          { label: '6', children: '', span: 1 },
        ]}
      />,
    );
    expect(container.querySelectorAll('.ant-descriptions-item')[0]).toHaveAttribute('colSpan', '2');
    expect(container.querySelectorAll('.ant-descriptions-item')[1]).toHaveAttribute('colSpan', '1');
    expect(container.querySelectorAll('.ant-descriptions-item')[2]).toHaveAttribute('colSpan', '1');
    expect(container.querySelectorAll('.ant-descriptions-item')[3]).toHaveAttribute('colSpan', '2');
    expect(container.querySelectorAll('.ant-descriptions-item')[4]).toHaveAttribute('colSpan', '3');
    expect(container.querySelectorAll('.ant-descriptions-item')[5]).toHaveAttribute('colSpan', '1');
    expect(container.querySelectorAll('.ant-descriptions-item')[6]).toHaveAttribute('colSpan', '2');
  });

  it('column is number', () => {
    const wrapper = render(
      <Descriptions column={3}>
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
    wrapper.unmount();
  });

  it('when typeof column is object', () => {
    const wrapper = render(
      <Descriptions column={{ xs: 8, sm: 16, md: 24 }}>
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
      </Descriptions>,
    );
    expect(
      Array.from(wrapper.container.querySelectorAll('td'))
        .map((i) => Number(i.getAttribute('colspan')))
        .filter(Boolean)
        .reduce((total, cur) => total + cur, 0),
    ).toBe(8);
    wrapper.unmount();
  });

  it('warning if exceed the row span', () => {
    resetWarned();

    render(
      <Descriptions column={3}>
        <Descriptions.Item label="Product" span={2}>
          Cloud Database
        </Descriptions.Item>
        <Descriptions.Item label="Billing" span={2}>
          Prepaid
        </Descriptions.Item>
      </Descriptions>,
    );
    expect(errorSpy).toHaveBeenCalledWith(
      'Warning: [antd: Descriptions] Sum of column `span` in a line not match `column` of Descriptions.',
    );
  });

  it('when item is rendered conditionally', () => {
    const hasDiscount = false;
    const wrapper = render(
      <Descriptions>
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
        {hasDiscount && <Descriptions.Item label="Discount">$20.00</Descriptions.Item>}
      </Descriptions>,
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
    wrapper.unmount();
  });

  it('vertical layout', () => {
    const wrapper = render(
      <Descriptions layout="vertical">
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
    wrapper.unmount();
  });

  it('Descriptions.Item support className', () => {
    const wrapper = render(
      <Descriptions>
        <Descriptions.Item label="Product" className="my-class">
          Cloud Database
        </Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });

  it('Descriptions support colon', () => {
    const wrapper = render(
      <Descriptions colon={false}>
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });

  it('Descriptions support style', () => {
    const wrapper = render(
      <Descriptions style={{ backgroundColor: '#e8e8e8' }}>
        <Descriptions.Item>Cloud Database</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });

  it('Descriptions support id', () => {
    const wrapper = render(
      <Descriptions id="descriptions">
        <Descriptions.Item>Cloud Database</Descriptions.Item>
      </Descriptions>,
    );
    const descriptionItemsElement = wrapper.container.querySelector('#descriptions');
    expect(descriptionItemsElement).not.toBeNull();
  });

  it('keep key', () => {
    render(
      <Descriptions>
        <Descriptions.Item key="bamboo">1</Descriptions.Item>
      </Descriptions>,
    );
    expect(jest.spyOn(document, 'createElement')).not.toHaveBeenCalled();
  });

  // https://github.com/ant-design/ant-design/issues/19887
  it('should work with React Fragment', () => {
    if (!React.Fragment) {
      return;
    }
    const wrapper = render(
      <Descriptions>
        <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
        <>
          <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
          <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
        </>
      </Descriptions>,
    );

    expect(wrapper.container.firstChild).toMatchSnapshot();
  });

  // https://github.com/ant-design/ant-design/issues/20255
  it('columns 5 with customize', () => {
    const wrapper = render(
      <Descriptions layout="vertical" column={4}>
        {/* 1 1 1 1 */}
        <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
        <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
        <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
        <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
        {/* 2 2 */}
        <Descriptions.Item label="bamboo" span={2}>
          bamboo
        </Descriptions.Item>
        <Descriptions.Item label="bamboo" span={2}>
          bamboo
        </Descriptions.Item>
        {/* 3 1 */}
        <Descriptions.Item label="bamboo" span={3}>
          bamboo
        </Descriptions.Item>
        <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
      </Descriptions>,
    );

    function matchSpan(rowIndex: number, spans: number[]) {
      const trs = Array.from(wrapper.container.querySelectorAll('tr')).at(rowIndex);
      const tds = Array.from(trs?.querySelectorAll('th')!);
      expect(tds).toHaveLength(spans.length);
      tds.forEach((td, index) => {
        expect(Number(td.getAttribute('colSpan'))).toEqual(spans[index]);
      });
    }

    matchSpan(0, [1, 1, 1, 1]);
    matchSpan(2, [2, 2]);
    matchSpan(4, [3, 1]);
  });

  it('number value should render correct', () => {
    const wrapper = render(
      <Descriptions bordered>
        <Descriptions.Item label={0}>0</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.container.querySelector('th')).toHaveClass('ant-descriptions-item-label');
    expect(wrapper.container.querySelector('td')).toHaveClass('ant-descriptions-item-content');
  });

  it('Descriptions support extra', () => {
    const wrapper1 = render(
      <Descriptions extra="Edit">
        <Descriptions.Item label="UserName">Zhou Maomao</Descriptions.Item>
      </Descriptions>,
    );
    const wrapper2 = render(
      <Descriptions>
        <Descriptions.Item label="UserName">Zhou Maomao</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper1.container.querySelector('.ant-descriptions-extra')).toBeTruthy();
    expect(wrapper2.container.querySelector('.ant-descriptions-extra')).toBeFalsy();
  });

  it('number 0 should render correct', () => {
    const wrapper = render(
      <Descriptions>
        <Descriptions.Item label={0} labelStyle={{ color: 'red' }} contentStyle={{ color: 'red' }}>
          {0}
        </Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });

  it('should pass data-* and accessibility attributes', () => {
    const { getByTestId } = render(
      <Descriptions data-testid="test-id" data-id="12345" aria-describedby="some-label">
        <Descriptions.Item label="banana">banana</Descriptions.Item>
      </Descriptions>,
    );
    const container = getByTestId('test-id');
    expect(container).toHaveAttribute('data-id', '12345');
    expect(container).toHaveAttribute('aria-describedby', 'some-label');
  });

  it('Descriptions should inherit the size from ConfigProvider if the componentSize is set', () => {
    const { container } = render(
      <ConfigProvider componentSize="small">
        <Descriptions bordered>
          <Descriptions.Item label="small">small</Descriptions.Item>
        </Descriptions>
      </ConfigProvider>,
    );
    expect(container.querySelectorAll('.ant-descriptions-small')).toHaveLength(1);
  });

  it('should items work', () => {
    const { container } = render(
      <Descriptions
        items={[
          {
            key: '1',
            label: 'UserName',
            children: 'Zhou Maomao',
          },
          {
            key: '2',
            label: 'Telephone',
            children: '1810000000',
          },
          {
            key: '3',
            label: 'Live',
            children: 'Hangzhou, Zhejiang',
          },
        ]}
      />,
    );
    expect(container.querySelector('.ant-descriptions-item')).toBeTruthy();
    expect(container.querySelectorAll('.ant-descriptions-item')).toHaveLength(3);
    expect(container).toMatchSnapshot();
  });

  it('Descriptions nested within an Item are unaffected by the external borderless style', () => {
    const { container } = render(
      <Descriptions bordered>
        <Descriptions.Item>
          <Descriptions bordered={false} />
        </Descriptions.Item>
      </Descriptions>,
    );

    const nestDesc = container.querySelectorAll('.ant-descriptions')[1];
    const view = nestDesc.querySelector('.ant-descriptions-view');
    expect(getComputedStyle(view!).border).toBeFalsy();
  });

  it('Should Descriptions not throw react key prop error in jsx mode', () => {
    render(
      <Descriptions title="User Info">
        <Descriptions.Item key="1" label="UserName">
          Zhou Maomao
        </Descriptions.Item>
        <Descriptions.Item label="Telephone">1810000000</Descriptions.Item>
      </Descriptions>,
    );
    expect(errorSpy).not.toHaveBeenCalledWith(
      expect.stringContaining('`key` is not a prop'),
      expect.anything(),
      expect.anything(),
    );
  });

  // https://github.com/ant-design/ant-design/issues/47151
  it('should has .ant-descriptions-item-content className when children is falsy', () => {
    const wrapper = render(
      <Descriptions
        bordered
        items={[
          {
            key: '1',
            label: null,
            children: null,
          },
        ]}
      />,
    );
    expect(wrapper.container.querySelectorAll('.ant-descriptions-item-label')).toHaveLength(1);
    expect(wrapper.container.querySelectorAll('.ant-descriptions-item-content')).toHaveLength(1);
  });
});
