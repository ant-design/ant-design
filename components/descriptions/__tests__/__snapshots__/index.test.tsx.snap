// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Descriptions Descriptions support colon 1`] = `
<div
  class="ant-descriptions"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label ant-descriptions-item-no-colon"
              >
                Product
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                Cloud Database
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions Descriptions support style 1`] = `
<div
  class="ant-descriptions"
  style="background-color: rgb(232, 232, 232);"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-content"
              >
                Cloud Database
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions Descriptions.Item support className 1`] = `
<div
  class="ant-descriptions"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item my-class"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                Product
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                Cloud Database
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions column is number 1`] = `
<div
  class="ant-descriptions"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                Product
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                Cloud Database
              </span>
            </div>
          </td>
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                Billing
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                Prepaid
              </span>
            </div>
          </td>
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                time
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                18:00:00
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="3"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                Amount
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                $80.00
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions number 0 should render correct 1`] = `
<div
  class="ant-descriptions"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
                style="color: red;"
              >
                0
              </span>
              <span
                class="ant-descriptions-item-content"
                style="color: red;"
              >
                0
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions should items work 1`] = `
<div>
  <div
    class="ant-descriptions"
  >
    <div
      class="ant-descriptions-view"
    >
      <table>
        <tbody>
          <tr
            class="ant-descriptions-row"
          >
            <td
              class="ant-descriptions-item"
              colspan="1"
            >
              <div
                class="ant-descriptions-item-container"
              >
                <span
                  class="ant-descriptions-item-label"
                >
                  UserName
                </span>
                <span
                  class="ant-descriptions-item-content"
                >
                  Zhou Maomao
                </span>
              </div>
            </td>
          </tr>
          <tr
            class="ant-descriptions-row"
          >
            <td
              class="ant-descriptions-item"
              colspan="1"
            >
              <div
                class="ant-descriptions-item-container"
              >
                <span
                  class="ant-descriptions-item-label"
                >
                  Telephone
                </span>
                <span
                  class="ant-descriptions-item-content"
                >
                  1810000000
                </span>
              </div>
            </td>
          </tr>
          <tr
            class="ant-descriptions-row"
          >
            <td
              class="ant-descriptions-item"
              colspan="1"
            >
              <div
                class="ant-descriptions-item-container"
              >
                <span
                  class="ant-descriptions-item-label"
                >
                  Live
                </span>
                <span
                  class="ant-descriptions-item-content"
                >
                  Hangzhou, Zhejiang
                </span>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Descriptions should work with React Fragment 1`] = `
<div
  class="ant-descriptions"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                bamboo
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                bamboo
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                bamboo
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                bamboo
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                bamboo
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                bamboo
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions vertical layout 1`] = `
<div
  class="ant-descriptions"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <th
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                Product
              </span>
            </div>
          </th>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-content"
              >
                Cloud Database
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <th
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                Billing
              </span>
            </div>
          </th>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-content"
              >
                Prepaid
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <th
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                time
              </span>
            </div>
          </th>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-content"
              >
                18:00:00
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <th
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                Amount
              </span>
            </div>
          </th>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-content"
              >
                $80.00
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions when item is rendered conditionally 1`] = `
<div
  class="ant-descriptions"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                Product
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                Cloud Database
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                Billing
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                Prepaid
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                time
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                18:00:00
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                Amount
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                $80.00
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
